### ISSUE TITLE
[Overlay][Windows8/7][Testpilot] Overlay icons are not shown

### ISSUE THREAD

**Issue by mcastroSG** [owncloud/client#5015]
### Expected behaviour
Overlay icons should appear on Shared sync folder

### Actual behaviour
Overlay icons are not show

### Steps to reproduce
1. Fresh install with [testpilotcloud-2.2.3.6210-nightly20160628-setup.exe](https://download.owncloud.com/desktop/daily/testpilotcloud-2.2.3.6210-nightly20160628-setup.exe)
2. Restart computer
3. Open Shared Syn folder

### Server configuration
ServerOS: Ubuntu 14.04 LTE
ServerVersion: 9.1.0 beta 2
### Client configuration

ClientOS: W7/W8.1
ClientVersion:  Version 2.2.3-nightly20160628 (build 6210)

Extra info:

 - Restart computer does not help this time
 - Regedit shows proper entries
 - ShExtView does not show anything abnormal
 - Date associated to shellext files 6/9/2016 6:31 PM (Seems to be the same files that official 2.2.2 client uses)

**Comment by guruz** [comment]
@mcastroSG But they work with the regular client? (non-branded)

**Comment by mcastroSG** [comment]
@guruz they are working fine on official patch 2.2.2 Build build 6192

**Comment by joel6890** [comment]
2.2.2, build 6192, Windows 10 Pro 64-bit: no overlay icons here but that's fine with me.

[Multiple folder copies problem](https://github.com/owncloud/client/issues/4329) has disappeared with this build, which is great. I'd rather have that problem gone than have the overlay icons work.

**Comment by mcastroSG** [comment]
@joel6890 have you tried to restart after upgrade Desktop Client ?. It works for me and overlay icons appear again 😄

**Comment by joel6890** [comment]
I rebooted; and now all is as it was before. Overlay icons are back, as is the strange behavior of many Documents folder shortcuts self-populating at random intervals in the left panel.

So I am no further ahead after the update then I was before. Shouldn't have rebooted! :smile:

**Comment by ArthurChiao** [comment]
hi all,
 
as i posted in [#4977](https://github.com/owncloud/client/issues/4977), when i right click a file in the sync dir, I could see some logs like this:

```shell
Line 398: 06-28 14:32:57:828 0x7ae160 OCC::SocketApi::slotNewConnection: SocketApi:  New connection QLocalSocket(0x8b1b608)
Line 398: 06-28 14:32:57:828 0x7ae160 OCC::SocketApi::slotNewConnection: SocketApi:  New connection QLocalSocket(0x8b1b608)
Line 399: 06-28 14:32:57:828 0x7ae160 OCC::SocketApi::sendMessage: SocketApi:  Sending message:  "REGISTER_PATH:D:\Users\zhao_yn\ownCloud"
Line 399: 06-28 14:32:57:828 0x7ae160 OCC::SocketApi::sendMessage: SocketApi:  Sending message:  "REGISTER_PATH:D:\Users\zhao_yn\ownCloud"
Line 400: 06-28 14:32:57:829 0x7ae160 OCC::SocketApi::sendMessage: SocketApi:  Sending message:  "SHARE_MENU_TITLE:Share with ownCloud"
Line 400: 06-28 14:32:57:829 0x7ae160 OCC::SocketApi::sendMessage: SocketApi:  Sending message:  "SHARE_MENU_TITLE:Share with ownCloud"
Line 401: 06-28 14:32:57:878 0x7ae160 OCC::SocketApi::onLostConnection: SocketApi:  Lost connection  QLocalSocket(0x8b1b608)
Line 401: 06-28 14:32:57:878 0x7ae160 OCC::SocketApi::onLostConnection: SocketApi:  Lost connection  QLocalSocket(0x8b1b608)
```

Can someone explain what's happening behind this scene? e.g. the server and client activities of the windows pipe. And why the connection lost immediately every time?

**Comment by guruz** [comment]
@dragotin ^^ isn't that what you've seen too on one of your machines? connection immediatly lost again?

@ArthurChiao The Explorer DLLs connect to the sync client via the pipe. They then exchange info about the file state for the Overlay icons. The 'Share with ownCloud' is also sent via the pipe.

**Comment by dragotin** [comment]
@guruz yes, exactly that.

**Comment by ArthurChiao** [comment]
@guruz I built the [owncloud-client-binary](https://github.com/owncloud/owncloud-client-binary) project with Visual Studio 2013, and am debugging into the DLLs.

There seems to be 2 phenomenons in my case:

1. `OCOverlays_x64.dll` does not send messages to client
1. `OCContextMenu_x64.dll` sends messages to client, and always gets (correct) response, but Explorer does nothing on receiving the response (should show `Share with ownCloud`, but is unstable, some times show and sometimes doesn't)

**Comment by ArthurChiao** [comment]
I added some log in `OCUtil/RemotePatchChecker.cpp`:

```cpp
                auto responseStatus = response.substr(statusBegin+1, statusEnd - statusBegin-1);
                auto responsePath = response.substr(statusEnd+1);
                auto state = _StrToFileState(responseStatus);
                bool wasAsked = asked.erase(responsePath) > 0;

                bool changed = false;
                {   std::unique_lock<std::mutex> lock(_mutex);
                    bool wasCached = _cache.find(responsePath) != _cache.end();
                    if (wasAsked || wasCached) {
                        auto &it = _cache[responsePath];
                        changed = (it != state);
                        it = state;
                    }
                }
                if (changed) {
                    SHChangeNotify(SHCNE_UPDATEITEM, SHCNF_PATH | SHCNF_FLUSHNOWAIT, responsePath.data(), NULL);
                }
				if (g_logfile.is_open())
				{
					g_logfile << "  reponseStatus: " << responseStatus << ", responsePath: " << responsePath << ", changed: " << changed << "\n";
					g_logfile.flush();
				}
```

here is the log when i add a new folder into my sync dir:

```
BROADCAST:   reponseStatus: SYNC, responsePath: D:\Users\zhao_yn\ownCloud\translations, changed: 0
BROADCAST:   reponseStatus: SYNC, responsePath: D:\Users\zhao_yn\ownCloud\translations\client_TW.ts, changed: 0
BROADCAST:   reponseStatus: SYNC, responsePath: D:\Users\zhao_yn\ownCloud\translations\client_th.ts, changed: 0
BROADCAST:   reponseStatus: SYNC, responsePath: D:\Users\zhao_yn\ownCloud\translations\client_tr.ts, changed: 0
BROADCAST:   reponseStatus: SYNC, responsePath: D:\Users\zhao_yn\ownCloud\translations\client_uk.ts, changed: 0
BROADCAST:   reponseStatus: SYNC, responsePath: D:\Users\zhao_yn\ownCloud\translations\client_zh_CN.ts, changed: 0
BROADCAST:   reponseStatus: SYNC, responsePath: D:\Users\zhao_yn\ownCloud\translations\client_zh_TW.ts, changed: 0
BROADCAST:   reponseStatus: SYNC, responsePath: D:\Users\zhao_yn\ownCloud\, changed: 0
BROADCAST:   reponseStatus: OK, responsePath: D:\Users\zhao_yn\ownCloud\translations, changed: 0
BROADCAST:   reponseStatus: OK, responsePath: D:\Users\zhao_yn\ownCloud\translations\client_TW.ts, changed: 0
BROADCAST:   reponseStatus: OK, responsePath: D:\Users\zhao_yn\ownCloud\translations\client_tr.ts, changed: 0
BROADCAST:   reponseStatus: OK, responsePath: D:\Users\zhao_yn\ownCloud\translations\client_th.ts, changed: 0
BROADCAST:   reponseStatus: OK, responsePath: D:\Users\zhao_yn\ownCloud\translations\client_zh_CN.ts, changed: 0
BROADCAST:   reponseStatus: OK, responsePath: D:\Users\zhao_yn\ownCloud\translations\client_uk.ts, changed: 0
BROADCAST:   reponseStatus: OK, responsePath: D:\Users\zhao_yn\ownCloud\translations\client_zh_TW.ts, changed: 0
BROADCAST:   reponseStatus: OK, responsePath: D:\Users\zhao_yn\ownCloud\translations, changed: 0
BROADCAST:   reponseStatus: OK, responsePath: D:\Users\zhao_yn\ownCloud, changed: 0
BROADCAST:   reponseStatus: OK, responsePath: D:\Users\zhao_yn\ownCloud\, changed: 0
BROADCAST:   reponseStatus: OK, responsePath: D:\Users\zhao_yn\ownCloud, changed: 0
UPDATE_VIEW
```

The code seems to say, if `changed` is `true`, it will change the folder status (**add overlay icon?**) by calling `SHChangeNotify`. In the log we can see the `changed` flag is always `false`, even if i sync a new folder with many files.

**And more, if i comment out the `if (changed)` clause, force it to call `SHChangeNotify`, i can not see overlay icons, still.**
```cpp
//if (changed) {
        SHChangeNotify(SHCNE_UPDATEITEM, SHCNF_PATH | SHCNF_FLUSHNOWAIT, responsePath.data(), NULL);
 //     }
```

I have followed [Overlay Icons WiKi](https://github.com/owncloud/client/wiki/Overlay-Icons) and [Debugging Overlay Icons](https://github.com/owncloud/client/wiki/Debugging-Overlay-Icons), checked my registry, it seems well registered.

**Comment by ArthurChiao** [comment]
updates:

`IFACEMETHODIMP OCOverlay::IsMemberOf(PCWSTR pwszPath, DWORD dwAttrib)` in `OCOverlay.cpp` seems **never** get called, whose return value determines whether or not to add overlay icon to a file.

Other methods, such as `IFACEMETHODIMP OCOverlay::QueryInterface(REFIID riid, void **ppv)` and `IFACEMETHODIMP OCOverlay::GetPriority(int *pPriority)`, are called correctly during initialization.

**Comment by ArthurChiao** [comment]
hi @guruz @mcastroSG  @dragotin @joel6890 , after massive debugging on both `client` and `OC*_x64.dll`s, i have fixed my problem.

The root cause lies in the client code, in [`socketapi.c`](https://github.com/owncloud/client/blob/2.2/src/gui/socketapi.cpp#L297) and [`folderman.cpp`](https://github.com/owncloud/client/blob/2.2/src/gui/folderman.cpp#L845): 

```cpp
Folder *FolderMan::folderForPath(const QString &path)
{
    QString absolutePath = QDir::cleanPath(path)+QLatin1Char('/');

    foreach(Folder* folder, this->map().values()) {
        const QString folderPath = folder->cleanPath()+QLatin1Char('/');

        if(absolutePath.startsWith(folderPath)) {
            //qDebug() << "found folder: " << folder->path() << " for " << absolutePath;
            return folder;
        }
    }

    return 0;
}
```

For each request from DLL, client first determines whether this folder exists on host, by calling `folderForPath()`. **However, the file path in request message is different from the client representation in terms of character cases:** for example, this is the debugging log i printed in `folderForPath()`. 

```
OCC::FolderMan::folderForPath: OCC::Folder* OCC::FolderMan::folderForPath(const QString&) , absolutePath:  "D:/Users/zhao_yn/cdrive/translations/" folderPath:  "d:/Users/zhao_yn/cdrive/"
```

This is the same folder on my computer, but the method returns false, then client sends `STATUS:NOP` to DLL, which means **this file does not exist on computer**, and DLL will not overlay any icon on receiving this message.

After changing the `if(absolutePath.startsWith(folderPath)) {` to `if(absolutePath.startsWith(folderPath, Qt::CaseInsensitive)) {`, everything turns OK.

**Besides, i also modified some `shell_integration/windows` code**,  such as, change [OCOverlay.cpp](https://github.com/owncloud/client/blob/master/shell_integration/windows/OCOverlays/OCOverlay.cpp#L107)
```cpp
IFACEMETHODIMP OCOverlay::GetPriority(int *pPriority)
{
	// this defines which handler has prededence, so
	// we order this in terms of likelyhood
	switch (_state) {
	case State_OK:
		*pPriority = 0;
	case State_OKShared:
		*pPriority = 1;
	case State_Warning:
		*pPriority = 2;
	case State_Sync:
		*pPriority = 3;
	case State_Error:
		*pPriority = 4;
	default:
		*pPriority = 5;
	}

	return S_OK;
}
```

to 
```cpp
IFACEMETHODIMP OCOverlay::GetPriority(int *pPriority)
{
	// this defines which handler has prededence, so
	// we order this in terms of likelyhood
	switch (_state) {
	case State_OK:
		*pPriority = 0; break;
	case State_OKShared:
		*pPriority = 1; break;
	case State_Warning:
		*pPriority = 2; break;
	case State_Sync:
		*pPriority = 3; break;
	case State_Error:
		*pPriority = 4; break;
	default:
		*pPriority = 5; break;
	}

	return S_OK;
}
```
I'm not sure whether it also helps, but i think the `breaks` are mandatory in logic.

BRs

**Comment by pierrejochem** [comment]
I had the same problem with a branded client. So this will solve it?

**Comment by pierrejochem** [comment]
Problems started with Version greater 2.2.0.0. Correct?

**Comment by ArthurChiao** [comment]
@pierrejochem yes, at least this solves my problem, and the code has been merged, see pull request #5024.

For more info, please also refer to #4977
